realm/realm-library/src/objectServer/java/io/realm/SyncSession.java

```diff
@@ -351,6 +414,56 @@ private void clearScheduledAccessTokenRefresh() {
         }
     }
 
+    // Wrapper class for handling the async operations of the underlying SyncSession calling `async_wait_for_download_completion`
+    private static class WaitForServerChangesWrapper {
+
+        private final CountDownLatch waitForChanges = new CountDownLatch(1);
+        private volatile boolean resultReceived = false;
+        private Long errorCode = null;
+        private String errorMessage;
+
+        /**
+         * Block until the wait either completes or is terminated for other reasons.
+         */
+        public void waitForServerChanges() {
+            if (!resultReceived) {
+                try {
+                    waitForChanges.await();
+                } catch (InterruptedException e) {
+                    RealmLog.warn("Downloading changes have been interrupted");
+                }
+            }
+        }
+
+        /**
+         * Process the result of a waiting action. This will also unblock anyone who called {@link #waitForChanges}.
+         *
+         * @param errorCode error code if an error occurred, {@code null} if changes were successfully downloaded.
+         * @param errorMessage error message (if any).
+         */
+        public void handleResult(Long errorCode, String errorMessage) {
+            this.errorCode = errorCode;
+            this.errorMessage = errorMessage;
+            this.resultReceived = true;
+            waitForChanges.countDown();
```

Comment:
I assume `handlerResult` will always be called even the sync session is closed. But if it is not, then it is more like something should be fixed in the ObjectStore/sync.